src/base/format_impl.h

```diff
@@ -0,0 +1,1088 @@
+		if (arg_index == 0) {
+			format_do_impl(out, orig_index, localize, t);
+			return false;
+		}
+		// By the way, this is not a recursive call – the compiler generates a separate
+		// function for every individual template specification, and they're all inlined.
+		return format_impl(out, orig_index, arg_index - 1, localize, args...);
+	}
+
+	explicit Tree(const char* format_string) : format_nodes_count_(0) {
+		for (; *format_string != '\0';) {
+			// read the next character and determine the kind of node we're looking at
+			if (*format_string != '%') {
+				// plain text, read until the next % sign
+				const char* start = format_string;
+				size_t len = 0;
+				for (; *format_string != '\0' && *format_string != '%'; ++format_string, ++len) {
+				}
+				nodes_.push_back(
+				   {std::unique_ptr<AbstractNode>(new LiteralNode(std::string(start, len))), 0});
+				continue;
+			}
+			// A format node. Let's see what it says.
+			++format_string;
+			if (*format_string == '\0') {
+				throw wexception("expected format sequence after '%%', found \\0");
+			}
+			if (*format_string == '%') {
+				nodes_.push_back({std::unique_ptr<AbstractNode>(new PercentNode), 0});
+				++format_string;
+				continue;
+			}
+
+			static unsigned format_index;
+			static unsigned precision;
+			static size_t min_width;
+			static uint8_t flags;
+			static bool min_w_found;
+			precision = kInfinitePrecision;
+			min_width = 0;
+			flags = kNone;
+			min_w_found = false;
+
+			if (*format_string >= '1' && *format_string <= '9') {
+				// Read a number. This is either a format index or the min width.
+				format_index = 0;
+				do {
+					format_index *= 10;
+					format_index += (*format_string - '0');
+					++format_string;
+				} while (*format_string >= '0' && *format_string <= '9');
+				if (*format_string == '\0') {
+					throw wexception("expected format sequence after '%%%u', found \\0", format_index);
+				}
+
+				if (*format_string == '%') {
+					// A node of the form '%N%' is complete.
+					AbstractNode* node = new WildcardNode;
+					nodes_.push_back({std::unique_ptr<AbstractNode>(node), format_index});
+					if (format_index > format_nodes_count_) {
+						format_nodes_count_ = format_index;
+						format_nodes_by_index_.resize(format_nodes_count_, nullptr);
+					}
+					if (format_nodes_by_index_[format_index - 1] != nullptr) {
+						throw wexception("duplicate use of index %u", format_index);
+					}
+					format_nodes_by_index_[format_index - 1] = node;
+					++format_string;
+					continue;
+				} else if (*format_string == '$') {
+					// $ character was recognized. Continue below.
+					++format_string;
+				} else {
+					// This was not a format index but the min_width.
+					// Correct the assumptions and jump forward.
+					min_width = format_index;
+					format_index = format_nodes_count_ + 1;
+					min_w_found = true;
+				}
+			} else {
+				format_index = format_nodes_count_ + 1;
+			}
```

Comment:
What happens, when the format string is like `"%3% %u %f"`? Doesn't `%3%` bump `format_nodes_count_` to 3, then `%u` and `%f` become 4 and 5? There's a check for missing indices in the end that will throw an exception in this case, but shouldn't mixing of numbered and unnumbered placeholders be forbidden in all cases to prevent mistakes like this?

The documentation above only says that "Argument indices are 1-based and there may be no gaps or duplicates. Unnumbered placeholders are enumerated from left to right." This could be interpreted as `"%3% %u %f"` is equivalent to `"%3% %1$u %2$f"`. This would need a search for the first unused index for unnumbered placeholders, wouldn't it?

Reply:
I added a check that positional and unpositional placeholders may not be mixed. Now either all placeholders should have a number, or none should have one.